### ISSUE TITLE
numba

### ISSUE THREAD

**Issue by irischang020** [chuckie82/skopi#4]
Tried to run the example script but got TypeError.
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-aa8f1eac6d96> in <module>
     38 tic = time.time()
     39 experiment = sk.SPIExperiment(det, beam, particle)
---> 40 img = experiment.generate_image()
     41 toc = time.time()
     42 print(">>> It took {:.2f} seconds to finish SPI calculation.".format(toc-tic))

~/skopi/skopi/experiment/base.py in generate_image(self, return_orientation)
     36             return self.det.assemble_image_stack(img_stack), orientation
     37         else:
---> 38             img_stack = self.generate_image_stack()
     39             return self.det.assemble_image_stack(img_stack)
     40 

~/skopi/skopi/experiment/base.py in generate_image_stack(self, return_photons, return_intensities, return_orientation, always_tuple)
     74             for i, particle_group in enumerate(sample_state):
     75                 group_complex_pattern += self._generate_group_complex_pattern(
---> 76                     recidet, i, particle_group)
     77 
     78             group_pattern = np.abs(group_complex_pattern)**2

~/skopi/skopi/experiment/base.py in _generate_group_complex_pattern(self, recidet, i, particle_group)
    108             pixel_momentum=recidet.pixel_position_reciprocal,
    109             orientations=orientations,
--> 110             inverse=True)
    111 
    112         for j, position in enumerate(positions):

~/skopi/skopi/geometry/slice_.py in take_n_slices(volume, voxel_length, pixel_momentum, orientations, inverse)
    124     for l in range(slice_num):
    125         slices_holder[l] = take_slice(volume, voxel_length, pixel_momentum,
--> 126                                       orientations[l], inverse)
    127 
    128     return slices_holder

~/skopi/skopi/geometry/slice_.py in take_slice(volume, voxel_length, pixel_momentum, orientation, inverse)
     75     """
     76     # construct the rotation matrix
---> 77     rot_mat = convert.quaternion2rot3d(orientation)
     78     if inverse:
     79         rot_mat = np.linalg.inv(rot_mat)

TypeError: expected dtype object, got 'numpy.dtype[float64]'
```
There seems to be issues with numba.
https://github.com/numba/numba/issues/6041

**Comment by chuckie82** [comment]
what conda env are you using?
pytest passes with `conda activate /cds/home/f/fpoitevi/miniconda3/envs/ana-py3`
could you do `conda list` to compare the versions? 
Maybe there's a minimum numba version needed for skopi...

**Comment by irischang020** [comment]
I followed Fred's instruction to build the environment: https://github.com/irischang020/ezpsana/tree/master/conda
Both environments (the one activated by calling `conda activate /cds/home/f/fpoitevi/miniconda3/envs/ana-py3` and the one I built) gives numba version 0.52.0
`numba                     0.52.0           py37hdc94413_0    conda-forge`

**Comment by fredericpoitevin** [comment]
We have been investigating and while we haven't solved Iris issue, lessons learned so far:
- `pip install -e .` your skopi in "my" environment seems to raise permission issues => better to build your own environment following instructions from `ezpsana`
- In the case of Iris, even after installing her own miniconda (i.e. not sourcing psconda.sh), and once in the newly built environment, `pip freeze | grep cupy` yields 2 `cupy`, one that starts with `cupy @ file:///home/conda...` and the other one that should not be there `cupy-cuda90-8.0.0rc1`

**Comment by fredericpoitevin** [comment]
So the solution was to do `pip uninstall cupy-cuda90`.

It is unclear to me why this version of `cupy` could be found in the environment, but that is one more reason in favor of conda within a container, instead of conda on bare metal. The latter will inherit from history thus potentially creating situations that are more difficult to debug.

**Comment by irischang020** [comment]
Thanks @fredericpoitevin for helping me with setting up the environment. I could `import skopi, cupy, psana` without problem now. However, there is still numba TypeError issue, which I think could be related to some package version incompatibility.

**Comment by fredericpoitevin** [comment]
hmmm can you copy-paste here what created the error and the error? Thanks!

**Comment by irischang020** [comment]
I tried to run the notebook with the code in ExampleSPI.py using (ana-cuda-py3). The error can be found in the first block of this thread.

**Comment by fredericpoitevin** [comment]
Can you point me to the notebook here? https://github.com/chuckie82/skopi/tree/main/examples/notebooks

So I can replicate the exact thing

**Comment by irischang020** [comment]
https://github.com/chuckie82/skopi/blob/main/examples/scripts/ExampleSPI.py

**Comment by chuckie82** [comment]
I have relaxed the requirement.txt in skopi. I'll try building skopi with same requirements.txt as pysingfel. If so, copying over requirements.txt from pysingfel before `pip install -e .` may help

**Comment by fredericpoitevin** [comment]
@irischang020 this is what I get if I don't set`USE_CUPY` to 1:

```bash
(ana-py3) [fpoitevi@psanagpu106 scripts]$ python ExampleSPI.py
BEFORE: # of photons per pulse = 1000000000000.0
>>> Increasing the number of photons per pulse by a factor 100.0
AFTER : # of photons per pulse = 100000000000000.0
photon energy = 4600.0 eV
BEFORE: detector distance = 0.5810000000000015 m
>>> Increasing the detector distance by a factor of 0.5
AFTER : detector distance = 0.29050000000000076 m
>>> It took 42.26 seconds to finish SPI calculation.
ExampleSPI.py:52: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()
```

and this is what I get after I do `export USE_CUPY=1`:
```bash
(ana-py3) [fpoitevi@psanagpu106 scripts]$ python ExampleSPI.py
BEFORE: # of photons per pulse = 1000000000000.0
>>> Increasing the number of photons per pulse by a factor 100.0
AFTER : # of photons per pulse = 100000000000000.0
photon energy = 4600.0 eV
BEFORE: detector distance = 0.5809999999999967 m
>>> Increasing the detector distance by a factor of 0.5
AFTER : detector distance = 0.29049999999999837 m
>>> It took 49.76 seconds to finish SPI calculation.
Traceback (most recent call last):
  File "ExampleSPI.py", line 50, in <module>
    viz = sk.Visualizer(experiment, diffraction_rings="auto", log_scale=True)
  File "/cds/home/f/fpoitevi/Toolkit/skopi/skopi/visualizer.py", line 28, in __init__
    -xp.min(recidet.pixel_position_reciprocal[..., 0]))))
  File "<__array_function__ internals>", line 6, in amin
  File "/cds/home/f/fpoitevi/miniconda3/envs/ana-py3/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2859, in amin
    keepdims=keepdims, initial=initial, where=where)
  File "/cds/home/f/fpoitevi/miniconda3/envs/ana-py3/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 87, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
  File "cupy/core/core.pyx", line 1188, in cupy.core.core.ndarray.__array__
TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

As far as I can tell, both "succeed" but have some trouble at the visualization part.

**Comment by fredericpoitevin** [comment]
Also, things look OK for me on JupyterHub:
![image](https://user-images.githubusercontent.com/4610338/108146484-2beddb80-7082-11eb-84bc-411f0a5da15d.png)

Did you update your ipykernel after successful build of ana-py3?

**Comment by irischang020** [comment]
I don't know why I still run into the issue..
```
(ana-py3) [psanagpu106:scripts]$ python ExampleSPI.py
BEFORE: # of photons per pulse = 1000000000000.0
>>> Increasing the number of photons per pulse by a factor 100.0
AFTER : # of photons per pulse = 100000000000000.0
photon energy = 4600.0 eV
BEFORE: detector distance = 0.5809999999999967 m
>>> Increasing the detector distance by a factor of 0.5
AFTER : detector distance = 0.29049999999999837 m
Traceback (most recent call last):
  File "ExampleSPI.py", line 47, in <module>
    img = experiment.generate_image()
  File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 38, in generate_image
    img_stack = self.generate_image_stack()
  File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 76, in generate_image_stack
    recidet, i, particle_group)
  File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 110, in _generate_group_complex_pattern
    inverse=True)
  File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 126, in take_n_slices
    orientations[l], inverse)
  File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 77, in take_slice
    rot_mat = convert.quaternion2rot3d(orientation)
TypeError: expected dtype object, got 'numpy.dtype[float64]'
(ana-py3) [psanagpu106:scripts]$ python
Python 3.7.9 | packaged by conda-forge | (default, Feb 13 2021, 20:03:11) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cupy as cp
>>> import skopi as sk
>>> import psana
```

**Comment by fredericpoitevin** [comment]
> I don't know why I still run into the issue..
> 
> ```
> (ana-py3) [psanagpu106:scripts]$ python ExampleSPI.py
> BEFORE: # of photons per pulse = 1000000000000.0
> >>> Increasing the number of photons per pulse by a factor 100.0
> AFTER : # of photons per pulse = 100000000000000.0
> photon energy = 4600.0 eV
> BEFORE: detector distance = 0.5809999999999967 m
> >>> Increasing the detector distance by a factor of 0.5
> AFTER : detector distance = 0.29049999999999837 m
> Traceback (most recent call last):
>   File "ExampleSPI.py", line 47, in <module>
>     img = experiment.generate_image()
>   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 38, in generate_image
>     img_stack = self.generate_image_stack()
>   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 76, in generate_image_stack
>     recidet, i, particle_group)
>   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 110, in _generate_group_complex_pattern
>     inverse=True)
>   File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 126, in take_n_slices
>     orientations[l], inverse)
>   File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 77, in take_slice
>     rot_mat = convert.quaternion2rot3d(orientation)
> TypeError: expected dtype object, got 'numpy.dtype[float64]'
> (ana-py3) [psanagpu106:scripts]$ python
> Python 3.7.9 | packaged by conda-forge | (default, Feb 13 2021, 20:03:11) 
> [GCC 9.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import cupy as cp
> >>> import skopi as sk
> >>> import psana
> ```

Is it with `$USE_CUPY` set to 1?

**Comment by irischang020** [comment]
> > I don't know why I still run into the issue..
> > ```
> > (ana-py3) [psanagpu106:scripts]$ python ExampleSPI.py
> > BEFORE: # of photons per pulse = 1000000000000.0
> > >>> Increasing the number of photons per pulse by a factor 100.0
> > AFTER : # of photons per pulse = 100000000000000.0
> > photon energy = 4600.0 eV
> > BEFORE: detector distance = 0.5809999999999967 m
> > >>> Increasing the detector distance by a factor of 0.5
> > AFTER : detector distance = 0.29049999999999837 m
> > Traceback (most recent call last):
> >   File "ExampleSPI.py", line 47, in <module>
> >     img = experiment.generate_image()
> >   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 38, in generate_image
> >     img_stack = self.generate_image_stack()
> >   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 76, in generate_image_stack
> >     recidet, i, particle_group)
> >   File "/cds/home/i/iris/skopi/skopi/experiment/base.py", line 110, in _generate_group_complex_pattern
> >     inverse=True)
> >   File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 126, in take_n_slices
> >     orientations[l], inverse)
> >   File "/cds/home/i/iris/skopi/skopi/geometry/slice_.py", line 77, in take_slice
> >     rot_mat = convert.quaternion2rot3d(orientation)
> > TypeError: expected dtype object, got 'numpy.dtype[float64]'
> > (ana-py3) [psanagpu106:scripts]$ python
> > Python 3.7.9 | packaged by conda-forge | (default, Feb 13 2021, 20:03:11) 
> > [GCC 9.3.0] on linux
> > Type "help", "copyright", "credits" or "license" for more information.
> > >>> import cupy as cp
> > >>> import skopi as sk
> > >>> import psana
> > ```
> 
> Is it with `$USE_CUPY` set to 1?

This is without `export USE_CUPY=1`.

**Comment by irischang020** [comment]
@fredericpoitevin comment out matplotlib.use('Agg') seems to solve the visualization issue. I am finally able to run the script on SDF using your (ana-py3), though still having trouble getting it to work on psana.

**Comment by fredericpoitevin** [comment]
So it seems that most issues are resolved with proper conda environments, however I still get a cupy related error.

When running the `virus_filling.ipynb` notebook in `skopi`'s examples (bypassing the cell with detector sign, since that cell is now obsolete - I will update the notebook soon) in the `ana-py3_latest.sif` container on SDF JupyterHub, I get to the following cell:
```python
experiment = sk.SPIExperiment(det, beam, particle)
image = experiment.generate_image()
#
fig = plt.figure(figsize=(4,3), dpi=180)
viz = sk.Visualizer(experiment, diffraction_rings="auto", log_scale=True)
viz.imshow(image)
```
- with `%env USE_CUPY=0`: 
![download](https://user-images.githubusercontent.com/4610338/108444065-c1b67180-720e-11eb-9ec1-576b32b7bbe4.png)
- with `%env USE_CUPY=1`:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-ab51e47df4c9> in <module>
      3 #
      4 fig = plt.figure(figsize=(4,3), dpi=180)
----> 5 viz = sk.Visualizer(experiment, diffraction_rings="auto", log_scale=True)
      6 viz.imshow(image)

~/skopi/skopi/visualizer.py in __init__(self, experiment, diffraction_rings, log_scale)
     26             -xp.min(recidet.pixel_position_reciprocal[..., 1]),
     27             xp.max(recidet.pixel_position_reciprocal[..., 0]),
---> 28             -xp.min(recidet.pixel_position_reciprocal[..., 0]))))
     29 
     30         self._auto_rings = False

<__array_function__ internals> in amin(*args, **kwargs)

/img/conda.local/miniconda3/envs/ana-py3/lib/python3.7/site-packages/numpy/core/fromnumeric.py in amin(a, axis, out, keepdims, initial, where)
   2857     """
   2858     return _wrapreduction(a, np.minimum, 'min', axis, None, out,
-> 2859                           keepdims=keepdims, initial=initial, where=where)
   2860 
   2861 

/img/conda.local/miniconda3/envs/ana-py3/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     85                 return reduction(axis=axis, out=out, **passkwargs)
     86 
---> 87     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
     88 
     89 

cupy/core/core.pyx in cupy.core.core.ndarray.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

**Comment by fredericpoitevin** [comment]
The weird visualization issues above (with and without cupy) are resolved in the PR "Cupy #6 " 
![image](https://user-images.githubusercontent.com/4610338/108803700-2ac02100-7550-11eb-8c93-7447bc344e78.png)
Also the SAXS class is now cupy-compatible. I think we can close this issue now.

**Comment by chuckie82** [comment]
great!
